### ISSUE TITLE
Filesize over 400kb. Is that expected? 

### ISSUE THREAD

**Issue by amir20** [react-bootstrap/react-bootstrap#1553]
Apologies if this question has been asked before.  I am using `react-bootstrap` for a side project. I noticed that uncompressed the js file is around `488KB` where the "jQuery" version of bootstrap.js is `69KB`. 

I found this by analyzing the bundle size. 

```
> $(npm bin)/webpack --config webpack-production.config.js --json | analyze-bundle-size
react: 576.93 kB (32.9%)
  <self>: 576.93 kB (100%)
react-bootstrap: 275.05 kB (15.7%)
  <self>: 275.05 kB (100%)
jquery: 241.79 kB (13.8%)
  <self>: 241.79 kB (100%)
immutable: 139.59 kB (7.96%)
  <self>: 139.59 kB (100%)
core-js: 125.89 kB (7.18%)
  <self>: 125.89 kB (100%)
moment: 101.96 kB (5.81%)
  <self>: 101.96 kB (100%)
lodash-compat: 67.98 kB (3.88%)
  <self>: 67.98 kB (100%)
react-overlays: 64.77 kB (3.69%)
  react-prop-types: 2.94 kB (4.53%)
    <self>: 2.94 kB (100%)
  react-prop-types/lib/elementType.js|/Users/amir.raminfar/Workspace/radiowitter: 1.11 kB (1.72%)
    <self>: 0 B (0.00%)
  <self>: 60.72 kB (93.7%)
fbjs: 32.87 kB (1.87%)
  <self>: 32.87 kB (100%)
babel-regenerator-runtime: 20.95 kB (1.19%)
...
```

I only use `Modal`s, `Col`s & `Row`s and `Button`s. 

Is this expected?

**Comment by jquense** [comment]
Well comparing the vanilla bootstrap js to the react-bootstrap file is misleading, since the react version also contains a ton of markup in the form of jsx, not to mention a bunch of simple components like Col, and Row that have no js equivalent in normal bootstrap.

That being said if you are only using 4 components make sure to only import those specific ones, by importing them directly from the `/lib` directort, see the example here: http://react-bootstrap.github.io/getting-started.html#commonjs

**Comment by amir20** [comment]
@jquense  I am importing them that way. See https://github.com/amir20/radiowitter/blob/master/app/assets/javascripts/components/controls.jsx#L3
`import { Button, Modal, Grid, Row, Col } from 'react-bootstrap'`

I don't think it is importing everything because using `analyze-bundle-stat` I see `react-bootstrap: 275.05 kB (15.7%)`. 

Isn't `275KB` for just `Button, Modal, Grid, Row, Col` still a lot? 

Also, I hear what you are saying about vanilla bootstrap vs this. But even after compressing, I get a ~88KB file for the `*.gz` file.

**Comment by taion** [comment]
488 kB is actually the size of our full, unminified bundle. Re-check your imports - you've most likely done something wrong.

**Comment by amir20** [comment]
No @taion. I didn't say my import is 400KB. My imported file size is actually actually 275KB. 

My question was why is the full size 400KB. React uncompressed is ~550KB. 

Please reopen as you are not understanding the problem.

**Comment by amir20** [comment]
I did some debugging I removed everything. That showed nothing from `react-bootstrap` being used. I then added only `import { Modal } from 'react-bootstrap'` and nothing else. 

Which yields: 
```
react: 576.93 kB (32.9%)
  <self>: 576.93 kB (100%)
react-bootstrap: 275.05 kB (15.7%)
  <self>: 275.05 kB (100%)
```

So does `<Modal/>` really need 275KB? I hope not.

**Comment by jquense** [comment]
> If you install React-Bootstrap using NPM, you can import individual components under react-bootstrap/lib rather than the entire library. Doing so pulls in only the specific components that you use, which will reduce the size of your client bundle.

@amir20 per the instructions on the site you are importing everything, not just the piece you want.

you need to import from `/lib` directly like: `import Button from 'react-bootstrap/lib/Button'`

**Comment by amir20** [comment]
According to documentation, `import Button from 'react-bootstrap/lib/Button'` is the same as import { Button  } from 'react-bootstrap'. But I tried both, and you were right. Then this makes more sense. 

Perhaps it's the way the project is structured?

**Comment by taion** [comment]
No that's literally just how CommonJS modules work.

**Comment by framerate** [comment]
@taion 

The internet is training people to do `import { Button } from 'react-bootstrap'` as common practice, but as mentioned here my bundle is 200kb bigger if I do that single line versus `import Button from 'react-bootstrap/lib/Button'`

What is the reasoning that they're not synonymous? Is it a webpack thing that should be intelligently importing the former to act like the latter or something that can be changed in react-bootstrap exports to make it happen?

**Comment by taion** [comment]
Again, that's literally just how CommonJS modules work. You can get a bit of relief with ES modules and tree-shaking if you're using webpack 2 or Rollup, but this isn't something we can control.

**Comment by framerate** [comment]
Can you elaborate a bit on ```that's literally just how CommonJS modules work```?

I was under the impression that ES6 modules (which I know have some inconsistencies in spec) are supposed to intelligently export when using the `import {something}` syntax

**Comment by taion** [comment]
@framerate See discussion https://github.com/reactjs/redux/pull/1369#issuecomment-210853615.

**Comment by framerate** [comment]
ahhh, thank you @taion !

**Comment by kirill-konshin** [comment]
@taion React Router has special folder `es6` for Webpack 2 and tree-shaking. Maybe you can support something similar?

**Comment by taion** [comment]
I added that build in React Router...

In practice tree-shaking is nearly useless for React Router and will be even closer to useless here.

**Comment by kirill-konshin** [comment]
Why useless? If someone imports like `import {Foo, Bar, Baz} from "react-bootstrap";` then tree-shaking should get rid of unused components, no?

**Comment by taion** [comment]
Sadly, no.

**Comment by kirill-konshin** [comment]
Hmm... then what tree-shaking actually does if not this? :)

**Comment by taion** [comment]
See discussion at https://github.com/reactjs/redux/pull/1369#issuecomment-210849415.

**Comment by kirill-konshin** [comment]
That makes sense, but still, in theory, if only buttons and popups were imported, the majority of other components should be gone, since those imports/exports are unused after all. Have you tried it? React Bootstrap is significantly bigger library than React Router, so maybe here it will be more useful.

**Comment by taion** [comment]
It doesn't matter what you import. Anything that can't be statically determined to lack side effects will be pulled in with ES6 imports from the top-level module. To ensure minimum bundle size, you _must_ import from the specific modules.

**Comment by Arturszott** [comment]
Hey! Sorry about digging in the old buried issues, yet I have one question connected to mentioned topic.  I solved imports mentioned in this issue by using: https://www.npmjs.com/package/babel-plugin-transform-imports yet React is still imported not in production version. Usually it should be like 100 KB, yet when I import some module from react-bootstrap react becomes 500KB package.

Any guess why it might be happening?

**Comment by Olivia101** [comment]
This issue can be solved by using [](https://www.npmjs.com/package/babel-plugin-transform-imports)

**Comment by zhenyulin** [comment]
Is there actually an even more elegant solution for this? It seems `dead_code: true` in webpack Uglify plugin doesn't remove those as unreachable codes, and it my bundle analyzer, multiple libraries are under similar situation, for instance `react-dom`

**Comment by jquense** [comment]
> Is there actually an even more elegant solution for this

Yes the recommendation in the documentation. Import only the files you are using. like `import Button from 'react-bootstrap/lib/Button'`

Also if you are not getting code removed, something is wrong with your uglify setup (I don't know what)

**Comment by zhenyulin** [comment]
I'm using multiple imports at the moment for multiple libraries, as it makes the code more concise, and currently manually configuring every lib in the `babel-plugin-transform-imports`, so actually was looking for a simpler way while maintain the usage of multiple imports

**Comment by jquense** [comment]
there isn't one that I know of. You pick the tradeoffs, personally I prefer explicit imports and more code

**Comment by sschenk** [comment]
@jquense 
> Yes the recommendation in the documentation. Import only the files you are using. like import Button from 'react-bootstrap/lib/Button'

I don't understand your answer. When I use:
`import Navbar from 'react-bootstrap/lib/Navbar'`

 I get undefined when I try to do <Navbar.Header />

**Comment by kirill-konshin** [comment]
It’s because you import default as Button, if you want to import all as button you need to write

‘’’js
import * as Navbar from 'react-bootstrap/lib/Navbar';
‘’’

**Comment by jquense** [comment]
You shouldn't need to do that. The sub components aren't named exports they are static properties on the default exported class
